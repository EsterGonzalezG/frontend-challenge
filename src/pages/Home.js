import React, { useEffect, useRef, useState } from 'react';
import { ItemList, localStorageList, Search } from './index';
export const Home = () => {
  const isMountedRef = useRef(true);
  const [valueSearch, setValueSearch] = useState('');
  const [list, setList] = useState({ data: [], loading: true });
  const [filterList, setFilterList] = useState([]);
  useEffect(() => {
    return () => {
      isMountedRef.current = false;
    };
  }, []);

  useEffect(() => {
    localStorageList().then((productList) => {
      if (isMountedRef.current) {
        setList({ data: productList, loading: false });
      }
    });
  }, []);

  useEffect(() => {
    setFilterList(
      list?.data?.filter(
        ({ brand, model }) =>
          brand.toLowerCase().includes(valueSearch.toLocaleLowerCase().trim()) ||
          model.toLowerCase().includes(valueSearch.toLocaleLowerCase().trim()),
      ),
    );
  }, [valueSearch, list.data]);

  return (
    <div className='wrapper'>
      <div className='l-content-wide'>
        <div className='l-paddingY-48'>
          <Search setValueSearch={setValueSearch} />
          <ItemList list={filterList} loading={list.loading} />
        </div>
      </div>
    </div>
  );
};
