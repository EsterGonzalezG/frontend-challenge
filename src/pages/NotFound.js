import React from 'react';
import { FaRegSadCry } from 'react-icons/fa';
import { ButtonBackToHome } from '../component/ButtonBackToHome';
export const NotFound = () => (
  <div className='notFount l-content-wide'>
    <span className='notFount-icon'>
      <FaRegSadCry />
    </span>
    <div className='notFount-text l-marginBottom-24'>UUUPPPSSS Sorry! This page is not available</div>
    <ButtonBackToHome>Home</ButtonBackToHome>
  </div>
);
