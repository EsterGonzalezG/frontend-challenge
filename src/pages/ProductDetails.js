import React, { useCallback, useContext, useEffect, useRef, useState } from 'react';
import { useLocation } from 'react-router-dom';
import {
  Actions,
  addProducts,
  CartContext,
  Description,
  getProductId,
  Image,
  NotData,
  setSessionStorage,
  Subtitle,
  Title,
} from './index';
const ProductDetails = () => {
  const query = new URLSearchParams(useLocation().search);
  const id = query.get('id');
  const isMountedRef = useRef(true);
  const { setCart } = useContext(CartContext);
  const [productDetail, setProductDetail] = useState({});
  const [colorCode, setColor] = useState('');
  const [storageCode, setStorage] = useState('');
  const { code, model, imgUrl } = productDetail;

  useEffect(() => {
    return () => {
      isMountedRef.current = false;
    };
  }, []);

  useEffect(() => {
    getProductId(id).then((data) => {
      if (isMountedRef.current) {
        setProductDetail(data);
      }
    });
  }, [id]);

  const addProductCart = useCallback(() => {
    addProducts({ id, colorCode, storageCode }).then(({ count }) => {
      setCart((current) => {
        setSessionStorage('cart', current + count);
        return current + count;
      });
    });
  }, [setCart, id, colorCode, storageCode]);

  if (code !== 0) {
    return (
      <div className='wrapper'>
        <div className='l-content-wide'>
          <div className='l-paddingY-24'>
            <Title>{model}</Title>
          </div>
          <div className='Card productDetails'>
            <div className='productDetails-img'>
              <Image image={imgUrl} model={model} />
            </div>
            <div className='productDetails-content'>
              <div className='productDetails-list'>
                <Subtitle>Characteristics:</Subtitle>
                <Description description={productDetail} />
              </div>
              <div className='l-paddingY-24'>
                <Subtitle>Select:</Subtitle>
                <Actions
                  product={productDetail}
                  setColor={setColor}
                  setStorage={setStorage}
                  addProductCart={addProductCart}
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  } else {
    return (
      <NotData button={true} textButton={'Home'}>
        Sorry, at the moment there is no data to display{' '}
      </NotData>
    );
  }
};
export default ProductDetails;
