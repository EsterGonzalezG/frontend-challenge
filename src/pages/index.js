import { Actions } from '../component/Actions';
import { ButtonBackToHome } from '../component/ButtonBackToHome';
import { Description } from '../component/Description';
import { Image } from '../component/Image';
import { ItemList } from '../component/ItemList';
import { addProducts } from '../services/addProducts';
import { getProductId } from '../services/getProducts';
import { CartContext } from '../useContext/CartContext';
import { localStorageList } from '../utils/localStorage';
import { NotData } from './../component/NotData';
import { Search } from './../component/Search';
import { Title } from './../component/Title';
import { setSessionStorage } from './../utils/sessionStorage';
export {
  ItemList,
  localStorageList,
  ButtonBackToHome,
  Actions,
  Description,
  Image,
  addProducts,
  getProductId,
  CartContext,
  NotData,
  Title,
  setSessionStorage,
  Search,
};
