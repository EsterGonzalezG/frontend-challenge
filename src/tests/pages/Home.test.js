import { mount } from 'enzyme';
import React from 'react';
import { Home } from './../../pages/Home';

describe('Test Home page', () => {
  const wrapper = mount(<Home />);
  test('should be render a component', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
