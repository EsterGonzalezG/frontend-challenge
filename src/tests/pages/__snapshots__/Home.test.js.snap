// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Home page should be render a component 1`] = `
<div
  className="wrapper"
>
  <div
    className="l-content-wide"
  >
    <div
      className="l-paddingY-48"
    >
      <form
        className="l-marginBottom-36 l-flex-justifyEnd"
        onSubmit={[Function]}
      >
        <label
          htmlFor="search"
        />
        <input
          className="Input"
          id="search"
          onChange={[Function]}
          placeholder="Mobile to search..."
          type="text"
        />
      </form>
      <div>
        LOADING...
      </div>
    </div>
  </div>
</div>
`;
